src/modsim2/similarity/metrics/kde.py

```diff
@@ -0,0 +1,295 @@
+
+        # Find the distances between samples in A and A
+        AA_distances, _ = A_neighbourhood.kneighbors(A, 2)
+        # Find the distances between samples in A and B
+        AB_distances, _ = B_neighbourhood.kneighbors(A, 1)
+
+        # Take the last distance for each
+        r = AA_distances[:, -1]
+        s = AB_distances[:, -1]
+
+        # Calculate distance according to corrected eqn 14 in paper
+        distance = np.log(s / r).sum() * d / n + np.log(m / (n - 1))
+
+        return distance
+
+    def calculate_distance(
+        self,
+        data_A: np.ndarray,
+        data_B: np.ndarray,
+        labels_A: np.ndarray,
+        labels_B: np.ndarray,
+        embedding_name: str,
+        kernel_name: str,
+        metric_name: str,
+        embedding_kwargs: dict = {},
+        integration_kwargs: dict = {},
+    ) -> float:
+        """
+        Calculates the distance between datasets A and B using kernel
+        density estimation to represent the probability density function
+        of the datasets.
+
+        The L2 norm and total variation are options that are provided
+        to calculate the distance between the distributions. The KL
+        divergence between the distributions can also be calculated;
+        however, this is not a true metric as it is not symmetrical.
+        An approximate method for calculating the KL divergence is
+        provided, although this does not require the density estimation.
+
+        Args:
+            data_A: The first image dataset
+            data_B: The second image dataset
+            labels_A: The target values for the first dataset
+            labels_B: The target values for the second dataset
+            embedding_name: The feature embeddings to use for the input arrays
+            kernel_name: The kernel to be applied in the density estimation, any
+                    kernels allowed in sklearns KernelDensity class are valid.
+                    Currently these are 'gaussian', 'tophat', 'epanechnikov',
+                    'exponential', 'linear', 'cosine'
+            metric_name: The metric to be used to calculate the distance between
+                    the density functions, currently only 'l2' is valid
+            embedding_kwargs: A dictionary of key word arguments to be passed
+                    to the embedding function
+            integration_kwargs: A dictionary of key word arguments to be passed
+                    to the integration function
+        """
+        # A value of zero is returned if the datasets are the same (the integration
+        # methods cannot be used if the datasets are the same, and the approximate
+        # kl method will calculate as -inf)
+        if np.array_equal(data_A, data_B):
+            return 0
+
+        # Check for valid embedding choice
+        assert embedding_name in EMBEDDING_FN_DICT, "Error: embedding does not exist"
+
+        if embedding_name in ["inception_umap"]:
```

Comment:
would prefer this was dealt with by umap function itself as a required arg unless this is a special case and we're happy with umap being non-deterministic elsewhere

Reply:
as discussed in meeting, this advice conflicts with previous advice to only do random seeding in once place! see if we can do this in original seeding location. Otherwise would still prefer this was done in umap rather than in kde (if can't be passed in from caller method)

Reply:
have removed this